### PR TITLE
fix(gc): classify boxed string concat as poll-capable

### LINKED ISSUE #7872: gc: js_string_concat_box is missing from POLL_CAPABLE_RUNTIME, blinding moving root-dominance arms
## Summary

The lightweight checker audit is red on current `main` (`79d2f59ed`):

```text
$ python3 scripts/gc_root_dominance_check.py --audit-poll-reach
=== poll reach: 3791 exported symbols, 1616 with an intra-runtime call edge, 384 matched by ALLOC_RE
error: ALLOC_RE symbols that CALL a POLL_CAPABLE_RUNTIME symbol but are not in POLL_CAPABLE_RUNTIME:
  js_string_concat_box                             -> js_dynamic_string_or_number_add
```

`js_string_concat_box` gained a call to poll-capable `js_dynamic_string_or_number_add` in `48c5e9a43` (#7835), but the caller was not added to `POLL_CAPABLE_RUNTIME`.

## Why this matters

The checker already classifies `js_string_concat_box` as a heap-value source via `ALLOC_RE`. Its body can now re-enter JS / reach a moving collection through `js_dynamic_string_or_number_add`, but a window whose only collection point is this helper classifies as `MOVING: no`.

That means **every gated `--moving-only` arm drops the window**, including the statepoint arm that checks the shipped lowering. This is the exact referent-with-no-name failure mode that #7616 and the `--audit-poll-reach` audit were added to prevent.

The workflow does run this audit, so current PR/main runs can report red, but `gc-root-dominance` is not a required branch-protection context (tracked separately in #7737 item 4).

Relevant source:

- [`js_string_concat_box` delegates non-string operands](https://github.com/PerryTS/perry/blob/79d2f59ed3a294c90c7ee52a5d5ffc9fa4fedcf7/crates/perry-runtime/src/string/concat.rs#L123-L146)
- [`--audit-poll-reach` explains the blind spot](https://github.com/PerryTS/perry/blob/79d2f59ed3a294c90c7ee52a5d5ffc9fa4fedcf7/scripts/gc_root_dominance_check.py#L782-L844)
- [The audit is wired before the expensive corpus build](https://github.com/PerryTS/perry/blob/79d2f59ed3a294c90c7ee52a5d5ffc9fa4fedcf7/.github/workflows/gc-root-dominance.yml#L151-L179)

## Acceptance criteria

- Add `js_string_concat_box` to `POLL_CAPABLE_RUNTIME` (or remove the poll-capable call only if semantics still match #7835).
- Add a focused checker self-test for an ALLOC_RE wrapper that delegates to a poll-capable helper, so this exact edge remains covered.
- `python3 scripts/gc_root_dominance_check.py --audit-poll-reach` passes.
- Re-run the moving/statepoint corpora and triage any newly exposed windows; do not merely make the structural audit green if adding the symbol reveals real root-dominance findings.

### PR DESCRIPTION
**Pull request by proggeramlug**
Closes #7872.

## Summary

- add `js_string_concat_box` to `POLL_CAPABLE_RUNTIME`, so every `--moving-only` checker arm sees the helper's path through `js_dynamic_string_or_number_add` to coercion/user code
- extend the planted poll-reach self-test with that exact two-hop production edge

## Reproduction

On the pre-fix `main` checkout:

```text
$ python3 scripts/gc_root_dominance_check.py --audit-poll-reach
error: ALLOC_RE symbols that CALL a POLL_CAPABLE_RUNTIME symbol but are not in POLL_CAPABLE_RUNTIME:
  js_string_concat_box                             -> js_dynamic_string_or_number_add
AUDIT_EXIT=2
```

After this change, the same audit reports 3,791 exported symbols, 1,616 intra-runtime call edges, 384 `ALLOC_RE` matches, and no unlisted poll-reaching allocation wrapper.

## Validation

- `python3 scripts/gc_root_dominance_check.py --self-test`
- all four static audits (`--audit-alloc-re`, `--audit-poll-capable`, `--audit-poll-reach`, `--audit-immovable-sources`)
- `python3 scripts/gc_gate_wiring_check.py`
- `python3 scripts/check_test_registration.py`
- shadow corpus: 139/139 sources, 161 IR modules, 0 skips
  - root-store dominance: 0 violations; 40/40 planted violations caught
  - unrooted allocas: 0
  - stale registers: 18, within the pinned 39 budget
- native corpus: 139/139 sources, 161 IR modules, 0 skips, 35,246 emitted statepoints / 21,290 non-empty live bundles
  - checked 35,085 safepoints and 50,900 relocations
  - 0 unrooted, 0 stale; 40/40 planted statepoint violations caught
- exact before/after classification over the curated corpus found 27 newly poll-reaching compiled functions but **0 newly exposed stale windows**
- `bash scripts/check_file_size.sh`
- `git diff --check`

Additional observation, not changed here: the dependency-scale Zod corpus emitted all 81 expected IR modules but its final link failed on unresolved intra-Zod symbols, identically with Perry's cache disabled. Running the checker directly over those emitted modules still met the normal floors (12,903 functions), reported 0 dominance violations, 0 unrooted allocas, and 89 stale uses within the 118 budget; the exact classification delta for this list entry was 0 newly exposed stale windows.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved garbage-collection safety checks for boxed string concatenation when coercion may execute user code.
  * Corrected poll-reachability analysis across indirect numeric coercion paths.
  * Added coverage for exact two-step coercion scenarios to improve audit accuracy and validation.

* **Documentation**
  * Added a changelog entry describing the corrected moving-only audit behavior and verification results.

### CODE REVIEW

**Comment by proggeramlug** [comment]
@coderabbitai review

File: changelog.d/7896-boxed-concat-poll-reach.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Fixed the GC root-dominance moving-only audits so boxed string concatenation is
+classified as poll-capable when it delegates non-string operands through
+coercion and user code.
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply changelog.d/7896-boxed-concat-poll-reach.md:3]: Addressed in 6e55f0623. The fragment now records the #7872 call-graph edge, the exact two-hop self-test, and the completed shadow/native corpus validation.
